Global rework of Postgresql url extension for Greenplum Database.

AndrewOvvv opened the pull request:
- Rewrite README.md with installation and usage guides
- Change installation for Greenplum, add functions into extension specific schema and change naming
- Rework source code
- Add more error reporting and comments in code

AndrewOvvv committed:
minor changes:
- change naming
- add and apply clang format
- change installation (create schema with functions) and set GP specific arguments
- add readme with installation giude and minor changes in usage

AndrewOvvv committed:
Allow to run functions on segments (in cases of merge)

AndrewOvvv committed:
rework encode function:
- shorter code
- use sprintf for easier printing '%XX' characters
- work with cstrings
also add some support functions and rework get_hex function:
- get_hex -> char2hex
- clear way to process errors:

AndrewOvvv committed:
apply clang-format to reworked code

AndrewOvvv committed:
global rework of decode function (support utf-16)

AndrewOvvv committed:
fix comment in control file

AndrewOvvv committed:
apply clang-format

AndrewOvvv committed:
changes in readme

robozmey reviewed (approved):
👍

noname0443 commented on src/gp_url_tools.c:
```diff
@@ -0,0 +1,229 @@
+}
+
+text *encode(text *input, const char *unreserved_special) {
+    int input_lenght;
```
I think there might be a typo with *`length`. Could you please amend it in other places in the code too?

AndrewOvvv replied:
You mean fix naming?

noname0443 replied:
Yes, `lenght` -> `length`.

AndrewOvvv replied:
Oh) I understand, my bad

noname0443 commented on src/gp_url_tools.c:
```diff
@@ -0,0 +1,229 @@
+static const unsigned int utf16_decode = 0x03FF;
+static const unsigned int utf16_decode_base = 0x10000;
+
+static const unsigned char hexlookup[256] = {
```
I believe using the array could be a bit confusing. It took me time to understand its functionality:
'0' - '9' => 0 - 9
'a' - 'f' => 10 - 15
'A' - 'F' => 10 - 15
Could we replace it with `if` statements in `char2hex` function? I think it would make it clearer.

noname0443 commented on src/gp_url_tools.c:
```diff
@@ -0,0 +1,229 @@
+            // special character => start process '%XX' or '%XXXX' sequence of
+            // chars
+            if (is_utf16(cinput + i, input_lenght - i)) {
+                // current sequence is in utf16 encoding
```
Doesn't the `if` statement above already account for the sequence being in UTF-16 encoding?

AndrewOvvv replied:
No, it can be utf-8 if sequence starts with '%'.

noname0443 replied:
It could be even so after we check it with `is_utf16` function? I meant that we apparently could remove the comment.

AndrewOvvv replied:
Okay, you issue about comment, I'll remove it

noname0443 replied:
Yes, github doesn't highlight precisely the line which I talk about :(

noname0443 commented on src/gp_url_tools.c:
```diff
@@ -0,0 +1,229 @@
+           sequence[1] != 'U';
+}
+
+bool is_utf16(const char *sequence, int lenght) {
```
May we add some format hints in comments? I believe they would make the code more reader-friendly.

AndrewOvvv replied:
I'll add it in next commit

AndrewOvvv committed:
fix issues from @noname0443

AndrewOvvv committed:
changes in error reporting, and more useful comments when error happens

AndrewOvvv committed:
Changes:
- add regression tests
- fix valid_utf16 function
- fix error with utf-16 decoding (not zeroed data in bytes and others local arrays

AndrewOvvv committed:
add asserts for better internal error detection and preventing segfaults
